### PR TITLE
Integrate feedback distribution within nbgrader

### LINKED ISSUE #795: Add feedback generation to the formgrader
Instructors should be able to run `nbgrader feedback` from the formgrader.

### PR DESCRIPTION
**Pull request by danielmaitre**
This is the result of the work of @dsblank, @perllaghu, @billywardrop29 and @danielmaitre at the Edinburgh hackathon.  

It includes UI integration of the feedback step, release of the feedback to students in the exchange, and fetching of this feedback by the students.

Some UI work needs to be finished but starting a PR will help check the coverage of the existing tests.

Fixes #795
Replaces #1096

### CODE REVIEW

**Comment by jhamrick** [comment]
Awesome, I will review this on the train 🙂 

Looks like there's a merge conflict since I merged #1119, so I think this will need to be rebased unfortunately.

**Comment by dsblank** [comment]
@danielmaitre I made a PR on yours that is a few steps closer on the submission display. I think there is a bug somewhere in fetch and display with proper available/view links. But I ran out of time today.

I can help refine all of this tomorrow on the plane home (I have a good 7 hours :) But maybe you will have solved all of the issues :)

File: nbgrader/exchange/releaseFeedback.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import os
```
Comment (jhamrick) [inline nbgrader/exchange/releaseFeedback.py]: Maybe this file could be called `release_feedback.py`, to be consistent with the name of the nbgrader command?

File: nbgrader/exchange/fetchfeedback.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import os
```
Comment (jhamrick) [inline nbgrader/exchange/fetchfeedback.py]: Maybe this file could be called `fetch_feedback.py`, to be consistent with the name of the nbgrader command?

File: nbgrader/exchange/releaseFeedback.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        self.course_path = os.path.join(self.root, self.course_id)
+        self.outbound_feedback_path = os.path.join(self.course_path, 'feedback')
+        self.dest_path = os.path.join(self.outbound_feedback_path)
+        # 0755
+        self.ensure_directory(
+            self.outbound_feedback_path,
+            S_IRUSR|S_IWUSR|S_IXUSR|S_IXGRP|S_IXOTH
+        )
+    # should be moved up to Exchange? Common to this and release
```
Comment (jhamrick) [inline nbgrader/exchange/releaseFeedback.py]: That would be great :)

**Comment by jhamrick** [comment]
P.S. I also think the docs need to be regenerated after the rebase, it looks like a bunch of the docs got removed so I think if you do a clean regeneration and commit the results the diff should be a bit smaller.

File: nbgrader/apps/api.py [hunk 4]
```diff
@@ -979,3 +979,104 @@ def autograde(self, assignment_id, student_id, force=True, create=True):
...
+            - error (string): formatted traceback
+            - log (string): captured log output
+
+        """
+        with temp_attrs(self.coursedir, assignment_id=assignment_id, student_id=student_id):
+            app = ExchangeFetchFeedback(coursedir=self.coursedir, parent=self)
+            ret_dic = capture_log(app)
+            # assignment tab needs a 'value' field with  
+        with temp_attrs(self.coursedir, assignment_id='*', student_id=student_id):
+            lister_rel = ExchangeList(inbound=False, coursedir=self.coursedir, config=self.config)
```
[the remark this answers is not in the document]
Comment (danielmaitre) [reply nbgrader/apps/api.py]: The buttons on the assignment_list tab require the POST to return an updated list of assignments. lister_sub and lister_rel seemed to return complementary lists to fill the two blocks on the assignment_list tab. When I left one out only one of the two blocks would fill.

File: nbgrader/apps/api.py [hunk 5]
```diff
@@ -979,3 +979,104 @@ def autograde(self, assignment_id, student_id, force=True, create=True):
...
+            - log (string): captured log output
+
+        """
+        if student_id is not None:
+            with temp_attrs(self.coursedir,
+                            assignment_id=assignment_id,
+                            student_id=student_id):
+                app = Feedback(coursedir=self.coursedir, parent=self)
+                c = Config()
+                c.HTMLExporter.template_file = 'feedback.tpl'
```
[the remark this answers is not in the document]
Comment (danielmaitre) [reply nbgrader/apps/api.py]: @dsblank should be able to comment on this one.

File: nbgrader/apps/api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jhamrick) [reply nbgrader/apps/api.py]: Ah ok yeah that makes sense, thanks! Maybe you could leave a comment in the code just explaining that?

[the remark this answers is not in the document]
Comment (dsblank) [reply nbgrader/apps/api.py]: I have a better version of this code in my PR.

[the remark this answers is not in the document]
Comment (danielmaitre) [reply nbgrader/apps/api.py]: I'll add one.

File: nbgrader/exchange/releaseFeedback.py (hunk 1 above)
Comment (danielmaitre) [reply nbgrader/exchange/releaseFeedback.py]: yes

File: nbgrader/exchange/fetchfeedback.py (hunk 2 above)
Comment (danielmaitre) [reply nbgrader/exchange/fetchfeedback.py]: yes

**Comment by danielmaitre** [comment]
Concerning the two identical submissions submitted twice but at different times, I personally prefer that they have the same exact feedback. The duplication happens often with my students when they are not sure the submission worked and they submit 5 times just to be sure. Probably the better solution would be to reject a submission (very politely) if it is identical to the previously submitted submission, this way we would get rid of virtually all duplicates?
--


"

File: nbgrader/exchange/releaseFeedback.py (hunk 3 above)
Comment (danielmaitre) [reply nbgrader/exchange/releaseFeedback.py]: Will do.

**Comment by jhamrick** [comment]
> the better solution would be to reject a submission (very politely) if it is identical to the previously submitted submission, this way we would get rid of virtually all duplicates?

I think that sounds like a good solution (though it would make testing a bit more annoying as it would require constantly clearing the exchange :) )

File: nbgrader/apps/api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsblank) [reply nbgrader/apps/api.py]: The issue here was that the HTMLExporter was being used in other place(s) in conversion. The code in the Feedback converter would only set the template to "feedback.tpl" if the template had not  already been set in the config. But it had been set, so we have to manually set it here to override that. (Hmmm, I wonder if we need to set it back...?)

File: nbgrader/apps/fetchfeedbackapp.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        """
+
+    @default("classes")
+    def _classes_default(self):
+        classes = super(FetchFeedbackApp, self)._classes_default()
+        classes.extend([Exchange, ExchangeFetchFeedback])
+        return classes
+
+    def _load_config(self, cfg, **kwargs):
+        if 'FetchApp' in cfg:
```
Comment (jhamrick) [inline nbgrader/apps/fetchfeedbackapp.py]: I don't think you need this here, as this is a deprecation warning for the regular (assignment) fetch functionality.

File: nbgrader/apps/releasefeedbackapp.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+        """
+    @default("classes")
+    def _classes_default(self):
+        classes = super(ReleaseFeedbackApp, self)._classes_default()
+        classes.extend([Exchange, ExchangeReleaseFeedback])
+        return classes
+
+    def _load_config(self, cfg, **kwargs):
+        if 'ReleaseFeedbackApp' in cfg:
```
Comment (jhamrick) [inline nbgrader/apps/releasefeedbackapp.py]: Same thing here, I don't think you need this warning.

File: nbgrader/apps/api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhamrick) [reply nbgrader/apps/api.py]: Thanks for clarifying---can you add a comment in the code with this explanation?

> I wonder if we need to set it back...?

I think probably not, since it creates a new config object on the fly here.

File: nbgrader/exchange/fetch_feedback.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        for nbname, timestamp, feedbackpath in self.feedbackFiles:
+            self.log.info( '{}'.format( (nbname, timestamp, feedbackpath) ) )
+            destWithTimestamp = os.path.join(dest,timestamp)    
+            if not os.path.isdir(destWithTimestamp):
+                os.makedirs(destWithTimestamp)
+            noExt, _ = os.path.splitext(nbname)
+            newName = noExt + '.html'
+            htmlFile = os.path.join(destWithTimestamp, newName)
+            shutil.copy(feedbackpath, htmlFile)
+        #else:
```
Comment (jhamrick) [inline nbgrader/exchange/fetch_feedback.py]: Can you remove this commented code?

File: nbgrader/nbextensions/assignment_list/assignment_list.js [hunk 9]
```diff
@@ -225,24 +225,92 @@ define([
...
+    AssignmentList.prototype.load_list_success = function (data, update_fetched_only) {
+	if (update_fetched_only == null) {
+            this.clear_list();
+	}
         var len = data.length;
+	var submissions = [];
         for (var i=0; i<len; i++) {
             var element = $('<div/>');
-            var item = new Assignment(element, data[i], this.fetched_selector, $.proxy(this.handle_load_list, this), this.options);
+            var item = new Assignment(element, submissions, data[i], this.fetched_selector,
```
Comment (jhamrick) [inline nbgrader/nbextensions/assignment_list/assignment_list.js]: Ah, I didn't realize it at first but I guess class populates `submissions`? That's a bit non-obvious---maybe leave a comment in the code saying that's what's happening.

Actually even better, if you don't mind making a few changes, would be not to pass `submissions` into the constructor but create it within the `Assignment` class and then below when you iterate over the submissions read it out as an attribute of the class.

File: nbgrader/docs/source/command_line_tools/index.rst [hunk 10]
```diff
@@ -21,7 +21,7 @@ Instructor commands
     nbgrader-validate
     nbgrader-update
     nbgrader-autograde
-    nbgrader-feedback
+    nbgrader-generate-feedback
```
Comment (jhamrick) [inline nbgrader/docs/source/command_line_tools/index.rst]: Is this correct? It doesn't look like you've renamed the command line app to `nbgrader generate_feedback` as far as I can tell---isn't it still `nbgrader feedback`?

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb [hunk 11]
```diff
@@ -1059,8 +1059,11 @@
    "source": [
     ".. seealso::\n",
     "\n",
-    "    :doc:`/command_line_tools/nbgrader-feedback`\n",
-    "        Command line options for ``nbgrader feedback``\n",
+    "    :doc:`/command_line_tools/nbgrader-generate-feedback`\n",
```
Comment (jhamrick) [inline nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]: Same comment here about `nbgrader feedback`

File: nbgrader/apps/api.py [hunk 12]
```diff
@@ -981,3 +981,104 @@ def autograde(self, assignment_id, student_id, force=True, create=True):
             app.force = force
             app.create_student = create
             return capture_log(app)
+
+    def generate_feedback(self, assignment_id, student_id=None):
+        """Run ``nbgrader generate_feedback`` for a particular assignment and student.
```
Comment (jhamrick) [inline nbgrader/apps/api.py]:
```suggestion
        """Run ``nbgrader feedback`` for a particular assignment and student.
```

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb [hunk 13]
```diff
@@ -1084,13 +1087,29 @@
...
     "\n",
-    "We can generate feedback based on these notebooks by running the `nbgrader feedback` command, which will produce HTML versions of these notebooks at the following location:\n",
+    "Creating feedback for students is divided into two parts:\n",
+    "\n",
+    "* generate feedback\n",
+    "* release feedback\n",
+    "\n",
+    "Generating feedback will create HTML files in the local instructor directory. Releasing feedback will copy those HTML files to the nbgrader exchange.\n",
+    "\n",
+    "We can generate feedback based on the graded notebooks by running the `nbgrader generate_feedback` command, which will produce HTML versions of these notebooks at the following location:\n",
```
Comment (jhamrick) [inline nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]:
```suggestion
    "We can generate feedback based on the graded notebooks by running the `nbgrader feedback` command, which will produce HTML versions of these notebooks at the following location:\n",
```

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb [hunk 14]
```diff
@@ -1084,13 +1087,29 @@
...
     "\n",
     "```\n",
     "feedback/{student_id}/{assignment_id}/{notebook_id}.html\n",
     "```\n",
     "\n",
-    "Unfortunately, `nbgrader feedback` is not yet available through the formgrader interface."
+    "The `nbgrader generate_feedback` is available by clicking the Generate Feedback button on either the Manage Assignments view (to generate feedback for all graded submissions), or on the individual student's Manage Submission page (to generate feedback for that specific individual)."
```
Comment (jhamrick) [inline nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]:
```suggestion
    "The `nbgrader feedback` command is available by clicking the Generate Feedback button on either the Manage Assignments view (to generate feedback for all graded submissions), or on the individual student's Manage Submission page (to generate feedback for that specific individual)."
```

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb [hunk 15]
```diff
@@ -1129,32 +1150,99 @@
...
-      "[FeedbackApp | INFO] Setting destination file permissions to 644\n"
+      "[NbGraderApp | WARNING] No nbgrader_config.py file found (rerun with --debug to see where nbgrader is looking)\n"
      ]
     }
    ],
    "source": [
     "%%bash\n",
     "\n",
-    "nbgrader feedback \"ps1\" --force"
+    "nbgrader generate_feedback \"ps1\" "
```
Comment (jhamrick) [inline nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]:
```suggestion
    "nbgrader feedback \"ps1\" "
```

File: nbgrader/docs/source/user_guide/highlights.rst [hunk 16]
```diff
@@ -42,3 +43,11 @@ commands:
 * `nbgrader release`: release a notebook to students
 * `nbgrader collect`: collect students' submissions
 * `nbgrader autograde`: autograde students' submissions
+* `nbgrader generate_feedback`: create feedback files from graded submissions
```
Comment (jhamrick) [inline nbgrader/docs/source/user_guide/highlights.rst]:
```suggestion
* `nbgrader feedback`: create feedback files from graded submissions
```

**Comment by dsblank** [comment]
@danielmaitre Unless you have already started on it, I'll make the submissions changes now as requested from Jess. After that, I'll look at the rest unless I hear from you. I have a few hours on the plane now. :)

**Comment by dsblank** [comment]
@danielmaitre Added the submissions changes as a PR on your repo: https://github.com/danielmaitre/nbgrader/pull/8

**Comment by dsblank** [comment]
@jhamrick @danielmaitre I intended for there to be three `nbgrader` commands:

* `nbgrader generate_feedback` (for instructor with exchange)
* `nbgrader release_feedback` (for instructor with exchange)
* `nbgrader fetch_feedback` (for student with exchange)

I think it would be confusing to call one of these `feedback`.

**Comment by dsblank** [comment]
(If we need to create an alias to `feedback` which we can deprecate one day, please let me know).

File: nbgrader/exchange/fetch_feedback.py (hunk 8 above)
Comment (dsblank) [reply nbgrader/exchange/fetch_feedback.py]: Done in https://github.com/danielmaitre/nbgrader/pull/8

File: nbgrader/nbextensions/assignment_list/assignment_list.js (hunk 9 above)
Comment (dsblank) [reply nbgrader/nbextensions/assignment_list/assignment_list.js]: Done in https://github.com/danielmaitre/nbgrader/pull/8

File: nbgrader/apps/fetchfeedbackapp.py (hunk 6 above)
Comment (dsblank) [reply nbgrader/apps/fetchfeedbackapp.py]: Removed method in https://github.com/danielmaitre/nbgrader/pull/8

File: nbgrader/apps/releasefeedbackapp.py (hunk 7 above)
Comment (dsblank) [reply nbgrader/apps/releasefeedbackapp.py]: Removed method in https://github.com/danielmaitre/nbgrader/pull/8

File: nbgrader/apps/api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsblank) [reply nbgrader/apps/api.py]: Added comments and refactor in https://github.com/danielmaitre/nbgrader/pull/8

**Comment by danielmaitre** [comment]
HI @dsblank , I won't look at this until tomorrow morning 6h30 am, so please make all the changes you want.
We broke the nbgrader/tests/nbextensions/test_assignment_list.py test. I think the problem is that none of the elements in the reworked submitted assignment list has a "list_item" class. We seem to have lost it on the way and it also causes the formatting to go weird.
Have a good flight!

**Comment by dsblank** [comment]
@danielmaitre Ok, sounds like a plan! I'll look at the bug above, and continue to make incremental fixes to the PR. The UI now looks like: 

Screenshot from 2019-06-02 09-04-26

which is similar to our original goals. I'm not sure I'll be able to do the expand/collapse because of updating the view when we click "Fetch Feedback"... exploring...

**Comment by jhamrick** [comment]
Cool, that sounds good! Re: `feedback` vs. `generate_feedback`, I am fine if it is `generate_feedback` but then there need to be a bunch of changes made as well to rename feedback to generate feedback. What might be easiest for now is to just keep the name `feedback` and in a separate PR rename `feedback` to `generate_feedback`, so this one doesn't get too big.

The screenshot looks nice! What does "available" mean though? That feedback is available to be fetched?

**Comment by jhamrick** [comment]
Another couple things (sorry I keep thinking of stuff)

1. You probably want to mark the tests you've added as `@notwindows` because the exchange functionality is not supported on windows.
2. You probably want to disable the "Release feedback" buttons in the UI if the exchange isn't available (which will be the case either on windows or when instructors haven't set up the exchange properly)

**Comment by dsblank** [comment]
Thanks @danielmaitre for fixing the tests to match the new submission UI.

Things left to do before merging:

- [ ] resolve conflicts
- [ ] mark tests notwindows
- [ ] hide ui parts when no exchange available
- [ ] change "available" in submission ui listing to "view" link
- [ ] don't change nbgrader command "release" to "generate_release" yet

I'll start on these (in no particular order) now unless I hear otherwise. 

Almost finished with this PR!

File: nbgrader/exchange/list.py [hunk 17]
```diff
@@ -79,20 +79,36 @@ def parse_assignments(self):
...
                     'notebook_id': os.path.splitext(os.path.split(notebook)[1])[0],
                     'path': os.path.abspath(notebook)
-                })
-
+                }
+                nb_hash = notebook_hash(notebook)
+                _, notebookFilename = os.path.split(notebook)
+                notebookName, _ = os.path.splitext(notebookFilename)
+                feedbackpath = os.path.join(self.root, info['course_id'], 'feedback', '{0}.html'.format(nb_hash))  
+                localFeedbackpath = os.path.join(self.coursedir.root, info['course_id'], 'feedback', '{0}.html'.format(notebookName))
```
Comment (dsblank) [inline nbgrader/exchange/list.py]: @danielmaitre I believe that the path assumes that the files are `course_id/feedback` but that should be either:

1. `course_id/assignment_id/feedback` (if organized by course), or
2. `assignment_id/feedback` (if not organized by course)

Comment (perllaghu) [reply nbgrader/exchange/list.py]: The current `fetch.py` code in the exchange does:

```python
        if self.path_includes_course:
            root = os.path.join(self.course_id, self.coursedir.assignment_id)
        else:
            root = self.coursedir.assignment_id
        self.dest_path = os.path.abspath(os.path.join('.', root))
```

Comment (danielmaitre) [reply nbgrader/exchange/list.py]: I'll fix the merge confilcts

Comment (danielmaitre) [reply nbgrader/exchange/list.py]: Yes the local path was not set properly, I wrote this bit before we decided what directory structure we wanted, and forgot to adapt it. I've now changed it and added the timestamp as well so now the UI shows "view" instead of available when the file is present.

Comment (danielmaitre) [reply nbgrader/exchange/list.py]: fixed a few more missing @notwindows

Comment (dsblank) [reply nbgrader/exchange/list.py]: Looks good! I'm turning the "view" into a link, and the "available" to "ready to fetch". Will make a PR within the next hour or so.

Comment (dsblank) [reply nbgrader/exchange/list.py]: PR adds link to "view", and changes text to "ready to fetch": https://github.com/danielmaitre/nbgrader/pull/9

**Comment by dsblank** [comment]
Here is the current state of the Submissions UI:

Screenshot from 2019-06-03 09-01-26

I guess we should remove the `Fetch Feedback` button if feedback has already been fetched for all submissions for that assignment. I originally left it there so students could easily re-fetch any feedback that had been re-released. But that is not consistent with the fetch and release of assignments.

**Comment by perllaghu** [comment]
Yep, remove it....
And Yes, once an assignment has been fetched [and exists on disk] it cannot
be re-fetched/updated

There is an argument that if there's feedback [for a submission] that's not
yet been fetched, and will therefore go into a directory that doesn't yet
exist, then there should be a Fetch button to get it

On Mon, 3 Jun 2019 at 14:04, Doug Blank <notifications@github.com> wrote:

> I guess we should remove the Fetch Feedback button if feedback has
> already been fetched for all submissions for that assignment. I originally
> left it there so students could easily re-fetch any feedback that had been
> re-released. But that is not consistent with the fetch and release of
> assignments.
>

-- 
Ian

**Comment by danielmaitre** [comment]
I've added an entry "allFeedbackDownloaded" in the assignment info passed to the UI along with "hasFeedback", and added logic to the UI to remove the button when there is nothing to download. Before this commit the CI tests were working, finger crossed they still will...

**Comment by dsblank** [comment]
I'll update the UI based on the new entry, and tie up the remaining loose ends in about 8 hours.

**Comment by perllaghu** [comment]
With the current exchange, I think you need to - when the exchange has a
proper API, then that logic [_allDownloaded_] should be moved into the
server_extension code

On Mon, 3 Jun 2019 at 15:36, danielmaitre <notifications@github.com> wrote:

> I've added an entry "allFeedbackDownloaded" in the assignment info passed
> to the UI along with "hasFeedback", and added logic to the UI to remove the
> button when there is nothing to download. Before this commit the CI tests
> were working, finger crossed they still will...
>

-- 
Ian

**Comment by jhamrick** [comment]
This is coming along really nicely 😄Let me know when you're ready for me to do a final pass!

File: nbgrader/docs/source/command_line_tools/index.rst (hunk 10 above)
Comment (dsblank) [reply nbgrader/docs/source/command_line_tools/index.rst]: Done

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb (hunk 11 above)
Comment (dsblank) [reply nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]: Done

File: nbgrader/apps/api.py (hunk 12 above)
Comment (dsblank) [reply nbgrader/apps/api.py]: Done

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb (hunk 13 above)
Comment (dsblank) [reply nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]: Done

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb (hunk 14 above)
Comment (dsblank) [reply nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]: Done

File: nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb (hunk 15 above)
Comment (dsblank) [reply nbgrader/docs/source/user_guide/creating_and_grading_assignments.ipynb]: Done

File: nbgrader/docs/source/user_guide/highlights.rst (hunk 16 above)
Comment (dsblank) [reply nbgrader/docs/source/user_guide/highlights.rst]: Done

**Comment by dsblank** [comment]
@danielmaitre One more PR: https://github.com/danielmaitre/nbgrader/pull/10

That addresses everything that I know about, I believe. Woot!

There is the artifact that if you submit an assignment more than once, then the feedback for one will be the feedback for all (and you'll see the "Fetch Feedback" button appear, and clicking it will download the old feedback for that one too). But I think we can address that by preventing submitting a notebook twice. In any event, we can refine that process after this PR.

@jhamrick Of course, after Daniel pulls 10, please let us know if there is anything you'd like changed.

**Comment by danielmaitre** [comment]
There are problems with the automatic CI tests, on travis only with python nighty and on appveyor only for python 3.5/6 (not sure which one it is, the executable says 3.5, the environment 3.6), anyone has any idea how to fix this?

**Comment by perllaghu** [comment]
@BertR - you did some stuff with the Travis at the hackathon, got time to have a look?

**Comment by dsblank** [comment]
Both errors on AppVeyor and Travis look like issues in nbconvert that has a missing language_info key in something. I don't think that is related to anything we (or nbgrader) has changed.

**Comment by jhamrick** [comment]
> Both errors on AppVeyor and Travis look like issues in nbconvert that has a missing language_info key in something. I don't think that is related to anything we (or nbgrader) has changed.

Yes, that is my suspicion as well, I have restarted the tests to see.

File: .gitignore [hunk 18]
```diff
@@ -112,3 +112,6 @@ package-lock.json
 .goutputstream*
 
 
+fakecoursedir/
```
Comment (jhamrick) [inline .gitignore]: Are these changes needed for this PR? They don't seem to actually appear in the other changes.

File: nbgrader/nbextensions/assignment_list/assignment_list.js [hunk 19]
```diff
@@ -315,14 +389,51 @@ define([
 
     Assignment.prototype.make_row = function () {
         var row = $('<div/>').addClass('col-md-12');
-        row.append(this.make_link());
+	var link = this.make_link();
```
Comment (jhamrick) [inline nbgrader/nbextensions/assignment_list/assignment_list.js]:
```suggestion
	 var link = this.make_link();
```

File: nbgrader/server_extensions/formgrader/apihandlers.py [hunk 20]
```diff
@@ -272,13 +273,54 @@ def post(self, assignment_id, student_id):
         self.write(json.dumps(self.api.autograde(assignment_id, student_id)))
 
 
+class GenerateAllFeedbackHandler(BaseApiHandler):
```
Comment (jhamrick) [inline nbgrader/server_extensions/formgrader/apihandlers.py]: Can you add the `@check_notebook_dir` decorator to all these new routes? (Sorry, it's from a PR I just merged recently)

File: .gitignore (hunk 18 above)
Comment (dsblank) [reply .gitignore]: Removed.

File: nbgrader/nbextensions/assignment_list/assignment_list.js (hunk 19 above)
Comment (dsblank) [reply nbgrader/nbextensions/assignment_list/assignment_list.js]: Done

File: nbgrader/server_extensions/formgrader/apihandlers.py (hunk 20 above)
Comment (dsblank) [reply nbgrader/server_extensions/formgrader/apihandlers.py]: Done

**Comment by dsblank** [comment]
@jhamrick @danielmaitre Final fixes can be found in https://github.com/danielmaitre/nbgrader/pull/11

**Comment by danielmaitre** [comment]
All checks have passed!

**Comment by perllaghu** [comment]
WOO-HOO!!!

Dang - that was some good work guys...

**Comment by dsblank** [comment]
I very much enjoyed working with you all! Great project to dive into most of the topics of nbgrader, and a great team! Let's do it again :)

**Comment by jhamrick** [comment]
Fabulous, thanks so much to all of you for working on this!

**Comment by dsblank** [comment]
Party!

Thanks for being our mentor through this @jhamrick ! I know how much patience it takes. Much appreciated!

**Comment by jhamrick** [comment]
I am more than happy to---it's really exciting seeing other want to contribute! 😄

**Comment by danielmaitre** [comment]
Thank you for your help on this project! 
Github says I can safely delete the branch, should I?

**Comment by jhamrick** [comment]
Fine by me.

**Comment by rkdarst** [comment]
I was looking at this, wanting to take it into our use.  It would be very nice to have and save us a lot of work!  But first I had to figure out how it works.  I was hoping that someone could confirm my observations below:

First, it seems to list the exchange inbound path (submitted assignments) to get notebook hashes, which are used to copy assignments back.  But the security of the inbound path is by not allowing students to list it.  Is this the case, do students have to list the inbound directory in order to fetch feedback?

Second, it also ensures that `{course_directory}/feedback/` is read/execute by students and fails with an error if not.  This would mean that students can list that entire directory and see all other students feedbacks.  Is this the expected situation?  At first I thought this could make the feedback dir read but not execute, so that students could get their feedback if they knew their exact filename (based on the hash), but not list all other assignments.

My thought, if the above is the case, chaneg it to do this: find the the hashes of the submitted notebooks from `~/.local/share/jupyter/nbgrader_cache/{course_id}/`, which only the submitter would have for their own notebooks.  Then, use that to get the feedback from the feedback dir, which they can access to copy known filenames from but not list.

**Comment by jhamrick** [comment]
I am not sure the answers to your questions off the top of my head, but can help try to answer them later this weekend 😄 in the meantime perhaps @danielmaitre @dsblank @perllaghu would know?

**Comment by Sefriol** [comment]
I mimicked how [Feedback Exchanger](https://github.com/jupyter/nbgrader/blob/master/nbgrader/exchange/fetch_feedback.py#L26) does determine submission and it's feedback in our environment.

By testing this with couple of submissions, you can indeed generate the same feedback md5-hash from `~/.local/share/jupyter/nbgrader_cache/{course_id}/` as you do from inbound directory which would guide the student to right feedback form. With this method student could then copy the feedback without giving student `ls` rights to the inbound directory. Much preferred way of doing things.

**Comment by jhamrick** [comment]
Let's continue discussion in #1202
